src/lib.rs

```diff
@@ -172,82 +172,57 @@
 //! [`Running`]: service::ServiceState::Running
 
 #![cfg(windows)]
-// Because of how deeply error-chain recurse with this many error types.
-#![recursion_limit = "128"]
 
 #[macro_use]
-extern crate error_chain;
+extern crate failure;
+pub type Result<T> = std::result::Result<T, failure::Error>;
```

Comment:
I'm still trying to wrap my head around how libraries and applications should best do error management. But with my current view on it I don't think libraries should ever return the `failure::Error` type. It becomes quite opaque to the library consumer, it forces allocation and dynamic dispatch upon them, and the returned error does not implement the `std::error::Error` trait. I kind of like how `failure` propose we should manage errors. But I don't like how it potentially divides the ecosystem into those who implement `std::error::Error` and those who implement `failure::Fail`.

As such I think applications can go full on with `failure`, but I think libraries should try to conform to the standard library style and return custom specific errors. I have recently started using the `err-derive` crate for libraries. It's the exact same thing as `#[derive(Fail)]` but it implements the `std::error::Error` trait instead. See here for an example: https://github.com/mullvad/hyper-dnscache/blob/master/src/cache_storer.rs#L132-L144

That library does only help with the creation of the custom error types. It does not include anything related to throwing errors or collecting backtraces etc. But it's still quite convenient IMO. Here is an example: https://github.com/mullvad/hyper-dnscache/blob/master/src/cache_storer.rs#L31-L33

As I said, this is something I'm still figuring out how to do best. So please give your feedback on what you think about this.

Reply:
I wouldn't say I have a well developed opinion on that matter. A while back we were working a couple PRs to get better legacy system support and it was mentioned switching to `failure` was something the project might be interested in. I won't be heart broken if that opinion has changed.

Reply:
I'm aware of the previous discussions. And I'm sorry I might have changed opinion since then. That was before I read the `failure` guide (https://rust-lang-nursery.github.io/failure/use-error.html). After doing so I was less convinced `failure` is a good fit for libraries. And especially the `Error` struct. The guide mentions the downsides I talk about and say that one use case for that struct is prototyping, which is certainly not what we do here now :)

I think that the goal of `failure` is to fix the standard library error handling. But until `std::error::Error` fully becomes the `Fail` trait I see it more as an ecosystem breakage. This is where `err-derive` comes in and picks out some of the best parts of `failure` but makes it std compliant.

Reply:
No need to be sorry, things change! I'm fine with closing this pr out and just count it as a learning experience.

Reply:
No need to close it really! We still want to get rid of `error-chain`. And using `err-derive` to define the error types is almost identical to how you do it for `failure`. Check their documentation. So if you have the energy you could look into migrating the PR to do that with fairly few changes I think. If not, I still want to thank you for the effort and I will look into this myself when I have time later.